packages/loader/container-loader/src/containerStorageAdapter.ts

```diff
@@ -126,12 +128,25 @@ export class ContainerStorageAdapter implements IDocumentStorageService, IDispos
 	}
 
 	public async getSnapshot(snapshotFetchOptions?: ISnapshotFetchOptions): Promise<ISnapshot> {
-		if (this._storageService.getSnapshot !== undefined) {
-			return this._storageService.getSnapshot(snapshotFetchOptions);
+		if (this._storageService.getSnapshot === undefined) {
+			throw new UsageError(
+				"getSnapshot api should exist in internal storage in ContainerStorageAdapter",
+			);
+		}
+
+		const key = snapshotFetchOptions?.loadingGroupIds?.sort().join();
+		const hasLoadingGroupIds = key !== undefined && key !== "";
+		// So when do we drop the this.loadingGroupIds cache? The problem is that we need it if we go offline.
+		const snapshot = hasLoadingGroupIds
+			? this.loadingGroupIds[key]
+			: await this._storageService.getSnapshot(snapshotFetchOptions);
+		if (
+			snapshotFetchOptions?.loadingGroupIds !== undefined &&
+			snapshotFetchOptions.loadingGroupIds.length > 0
+		) {
+			this.loadingGroupIds[snapshotFetchOptions.loadingGroupIds.sort().join()] = snapshot;
```

Comment:
I do not understand the logic with join & sort (hear and earlier in PR). I think it's the issue that you are describing in previous comment. Either way, it does not look right to me

Reply:
I made this clearer now, now it stores a snapshot if the network snapshot is newer or if there isn't one in the first place.

Reply:
And it stores a snapshot for each groupId